spot-contracts/contracts/_utils/BondHelpers.sol

```diff
@@ -191,27 +191,33 @@ library BondHelpers {
         return (td, collateralBalances, trancheSupplies);
     }
 
-    /// @notice Given a bond, retrieves the collateral redeemable for
-    ///         each tranche held by the given address.
+    /// @notice Given a bond and a user address computes the tranche amounts proportional to the tranche ratios,
+    //          that can be redeemed for the collateral before maturity.
     /// @param b The address of the bond contract.
     /// @param u The address to check balance for.
-    /// @return The tranche data and an array of collateral balances.
-    function getTrancheCollateralBalances(IBondController b, address u)
+    /// @return The tranche data and an array of tranche token balances.
+    function computeRedeemableTrancheAmounts(IBondController b, address u)
         internal
         view
         returns (TrancheData memory, uint256[] memory)
     {
-        TrancheData memory td;
-        uint256[] memory collateralBalances;
-        uint256[] memory trancheSupplies;
-
-        (td, collateralBalances, trancheSupplies) = getTrancheCollateralizations(b);
+        TrancheData memory td = getTrancheData(b);
+        uint256[] memory redeemableAmts = new uint256[](td.trancheCount);
+
+        // We calculate the minimum value of {trancheBal/trancheRatio} across tranches
+        uint256 min = type(uint256).max;
+        uint8 i;
+        for (i = 0; i < td.trancheCount && min != 0; i++) {
+            uint256 d = (td.tranches[i].balanceOf(u) * TRANCHE_RATIO_GRANULARITY) / td.trancheRatios[i];
+            if (d < min) {
+                min = d;
+            }
+        }
 
-        uint256[] memory balances = new uint256[](td.trancheCount);
-        for (uint8 i = 0; i < td.trancheCount; i++) {
-            balances[i] = (td.tranches[i].balanceOf(u) * collateralBalances[i]) / trancheSupplies[i];
+        for (i = 0; i < td.trancheCount; i++) {
+            redeemableAmts[i] = (td.trancheRatios[i] * min) / TRANCHE_RATIO_GRANULARITY;
```

Comment:
I'm confused by this section. I thought it would be the balances of tranches that can be redeemed. But I don't see the tranche balances of the users referenced anywhere here.

Also, wouldn't there be a different scalar for each tranche?

Say there's a bond that's configured as [20, 30, 50]
User's balances are [20, 25, 5000], so the limiting factor is tranche[1]
So it should return [16, 25, 40]

I would expect that we would need a different scalar number for each tranche balances. These would be 1x for the limiter and <= 1 for everything else. But it looks like this uses the same scalar for all of them.

Reply:
maybe the variables can be named differently

Consider A,B,Z with 0.2,0.3 and 0.5 and balances [10, 50, 20]

step one: we calculate balance/ratio -> 10/0.2, 50/0.3, 20/0.5 -> 50, 166.66, 40
step two: chose min (d=40)
step three: get amounts 40*0.2, 40*0.3, 40*0.5 -> [8, 12, 20]

d is essentially the maximum AMPL we can "re-tranche" from the existing balances

Reply:
Ah ok,  yeah that makes sense. Code comments could clear that further. What about:

At top of loop-- ```Calculate how many underlying assets could be redeemed from each tranche balance, assuming other tranches are not an issue, and record the smallest amount.```

Then maybe rename `min` -> `minUnderlyingOut`